include/boost/graph/link_cut_tree.hpp

```diff
@@ -0,0 +1,279 @@
+        inline Element find_path(Element x)
+        {
+            splay(x);
+            return x;
+        }
+
+        template <class Element>
+        inline Element find_tail(Element x)
+        {
+            while (get(right, x) != x)
+                x = get(right, x);
+            splay(x);
+            return x;
+        }
+
+        template <class Element>
+        inline Element join(Element u, Element v, Element w)
+        {
+            if (u != v)
+                put(parent, u, v);
+            if (w != v)
+                put(parent, w, v);
+            put(left, v, u);
+            put(right, v, w);
+            return v;
+        }
+
+        template <class Element>
+        inline std::pair<Element, Element> split(Element x)
+        {
+            splay(x);
+            Element x_left = get(left, x);
+            Element x_right = get(right, x);
+            if (x_left != x)
+                put(parent, x_left, x_left);
+            if (x_right != x)
+                put(parent, x_right, x_right);
+            put(left, x, x);
+            put(right, x, x);
+            return std::make_pair<>(x_left, x_right);
+        }
+    };
+    
+    
+    template <class ID = identity_property_map>
+    class link_cut_tree_with_storage :
+    public link_cut_tree<associative_property_map<std::unordered_map<typename property_traits<ID>::value_type, typename property_traits<ID>::value_type> > >
+    {
+    public:
+        typedef typename property_traits<ID>::key_type Vertex;
+        typedef typename property_traits<ID>::value_type Index;
+        typedef std::unordered_map<Index, Index> IndexMapContainer;
+        typedef associative_property_map<IndexMapContainer> IndexMap;
+        typedef link_cut_tree<IndexMap> LCT;
+        
+        link_cut_tree_with_storage(ID id_ = ID()) :
+        LCT(IndexMap(parent_map), IndexMap(left_map), IndexMap(right_map)),
+        id(id_) {}
+        
+        template <class Vertex>
+        inline void make_tree(Vertex x)
+        {
+            const Index x_id = get(id, x);
+            LCT::make_tree(x_id);
+            id_to_vertex[x_id] = x;
+        }
+        
+        template <class Vertex>
+        inline Vertex find_root(Vertex x)
+        {
+            return id_to_vertex[LCT::find_root(get(id, x))];
+        }
+        
+        template <class Vertex>
+        inline void link(Vertex x, Vertex y)
+        {
+            LCT::link(get(id, x), get(id, y));
+        }
+        
+        template <class Vertex>
+        inline void cut(Vertex x)
+        {
+            LCT::cut(get(id, x));
+        }
+        
+        template <class Vertex>
+        inline Vertex lowest_common_ancestor(Vertex x, Vertex y)
+        {
+            return id_to_vertex[LCT::lowest_common_ancestor(get(id, x), get(id, y))];
+        }
+        
+    private:
+        ID id;
+        std::unordered_map<Index, Vertex> id_to_vertex;
+        IndexMapContainer parent_map, left_map, right_map;
+    };
```

Comment:
Did you consider parameterizing the associative container for `id_to_vertex`? Someone might have a use case where they can just use `std::vector` and they'll be awfully annoyed that `std::unordered_map` is hard-coded.

Reply:
Thanks for advice, it's fixed. Also added documentation :)